### PR TITLE
FIX: Improve scanning and error handling under different scenarios.

### PR DESCRIPTION
**Pull request by romanrizzi**
This PR packs a bunch of fixes and improvements to how we scan uploads and handle different errors. We logged unhandled exceptions in the ClamAV client class, which helped us identify these issues. I'll break them down one by one:

---

`undefined method `read' for nil:NilClass while data = file.read(2048)`

Our store may return nil if it fails to download the upload from the store. Mark the scanned upload as failed with a custom error message and retry in 6hs.

`no implicit conversion of nil into String` - `Broken pipe - send(2)` - `Connection reset by peer - send(2)`

From the [ClamAV docs](https://linux.die.net/man/8/clamd):

```
Clamd requires clients to read all the replies it sent, before sending more commands to prevent send() deadlocks.

[...]

If clamd detects that a client has deadlocked, it will close the connection. Note that clamd may close an IDSESSION connection too if you don't follow the protocol's requirements. The client can use the PING command to keep the connection alive.
```

We read the ClamAV response from the socket using `getc`, which can return `nil`. To mitigate this issue, we use `IO#select` with a timeout to wait until we can read from the socket instead of doing it inmediately after streaming the file.

Since we receive `nil` and left unread messages, ClamAV thinks we deadlocked, resetting the connection when we try to scan the next upload.

Finally, we also need to handle when they send us an error response, so if the result contains the substring `ERROR`, we assume the scan failed and retry again in 6 hours. One of the errors is because the file is too large, and it turns out we are scanning data exports sent by system. We should skip those, so I tweaked the `BackgroundScan#queue_batch` query to reflect it.

### CODE REVIEW

**Review by pmusaraj — APPROVED** [review]
Looks good, thanks for the detailed explanation in the PR body.

File: lib/discourse_antivirus/background_scan.rb [hunk 1]
```diff
@@ -28,9 +28,15 @@ def stats
 
     def queue_batch(batch_size: 1000)
       Upload
+        .distinct
```
Comment (ZogStriP) [inline lib/discourse_antivirus/background_scan.rb:31]: Just curious: why is this needed? You seem to be doing only `left joins` which won't duplicate rows? 🤔

Comment (romanrizzi) [reply lib/discourse_antivirus/background_scan.rb:31]: `left joins` could still create duplicates if it's a 1:N relation, no? I assumed `upload_references` and `posts` could have multiple references to the same upload.
